client/core/exchangeratefetcher.go

```diff
@@ -0,0 +1,270 @@
+	DefaultFiatCurrency = "USD"
+	// fiatRateRequestInterval is the amount of time between calls to the exchange API.
+	fiatRateRequestInterval = 5 * time.Minute
+	// fiatRateDataExpiry : Any data older than fiatRateDataExpiry will be discarded.
+	fiatRateDataExpiry = 60 * time.Minute
+
+	// Tokens. Used to identify fiat rate source.
+	messari       = "Messari"
+	coinpaprika   = "Coinpaprika"
+	dcrdataDotOrg = "dcrdata"
+)
+
+var (
+	dcrDataURL     = "https://explorer.dcrdata.org/api/exchangerate"
+	coinpaprikaURL = "https://api.coinpaprika.com/v1/tickers/%s"
+	messariURL     = "https://data.messari.io/api/v1/assets/%s/metrics/market-data"
+	btcBipId, _    = dex.BipSymbolID("btc")
+	dcrBipId, _    = dex.BipSymbolID("dcr")
+)
+
+// fiatRateFetchers is the list of all supported fiat rate fetchers.
+var fiatRateFetchers = map[string]rateFetcher{
+	coinpaprika:   fetchCoinpaprikaRates,
+	dcrdataDotOrg: fetchDcrdataRates,
+	messari:       fetchMessariRates,
+}
+
+// fiatRateInfo holds the fiat rate and the last update time for an
+// asset.
+type fiatRateInfo struct {
+	rate       float64
+	lastUpdate time.Time
+}
+
+// rateFetcher can fetch fiat rates for assets from an API.
+type rateFetcher func(context context.Context, logger dex.Logger, assets map[uint32]*SupportedAsset) map[uint32]float64
+
+type commonRateSource struct {
+	mtx         sync.RWMutex
+	lastRequest time.Time
+	log         dex.Logger
+	fetchRates  rateFetcher
+	fiatRates   map[uint32]*fiatRateInfo
+}
+
+// logRequest sets the lastRequest time.Time.
+func (source *commonRateSource) logRequest() {
+	source.mtx.Lock()
+	defer source.mtx.Unlock()
+	source.lastRequest = time.Now()
+}
+
+// lastTry is the last time the fiat rate fetcher made a request.
+func (source *commonRateSource) lastTry() time.Time {
+	source.mtx.RLock()
+	defer source.mtx.RUnlock()
+	return source.lastRequest
+}
+
+// isExpired checks the last update time for all fiat rates against the
+// provided expiryTime.
+func (source *commonRateSource) isExpired(expiryTime time.Duration) bool {
+	source.mtx.Lock()
+	defer source.mtx.Unlock()
+	var expiredCount int
+	for _, rateInfo := range source.fiatRates {
+		if time.Since(rateInfo.lastUpdate) > expiryTime {
+			expiredCount++
+		}
+	}
+	totalFiatRate := len(source.fiatRates)
+	return expiredCount == totalFiatRate && totalFiatRate != 0
+}
+
+// assetRate returns the fiat rate information for the assetID specified.
+func (source *commonRateSource) assetRate(assetID uint32) *fiatRateInfo {
+	source.mtx.Lock()
+	defer source.mtx.Unlock()
+	rateInfo := source.fiatRates[assetID]
+	var fiatRateInfo fiatRateInfo
+	if rateInfo != nil {
+		fiatRateInfo = *rateInfo
+	}
+	return &fiatRateInfo
+}
+
+// refreshRates updates the last update time and the rate information for assets.
+func (source *commonRateSource) refreshRates(ctx context.Context, assets map[uint32]*SupportedAsset) {
+	source.logRequest()
+	fiatRates := source.fetchRates(ctx, source.log, assets)
```

Comment:
In general, I strongly dislike giving level structs like `commonRateSource` a log field.  If it's inconvenient to pass errors to the caller, we can just as easily pass the logger with `ctx` when calling `refreshRates` (just like the individual `rateFetcher`s).